question-generation/nlp.py

```diff
@@ -0,0 +1,42 @@
+from transformers import pipeline
+
+
+def interview_ai(sequence: str, num_questions: int) -> "list[str]":
+    """Generate questions based on a given sequence of context using Interview AI.
+
+    Args:
+        sequence: A string representing the context from which to generate questions.
+        num_questions: The number of questions to generate.
+
+    Returns:
+        A list of generated questions as strings.
+    """
+
+    model_id = "hyechanjun/interview-question-remake"
+    pipe = pipeline("text2text-generation", model=model_id, device=0)
```

Comment:
The pipeline should be loaded once and reused. That's the point of it being an object.

Reply:
It seems like you’re suggesting moving this code to the global space for reusability. However, I’ve already asked Ray to implement the ChatGPT method in a similar style, with the same input and output types as this function. The goal is to collect different NLP methods in this module and have `get_questions` function choose which one to use. Do you think it would be better to not have this as a function at all? My intention was to avoid pre-loading pipelines that wouldn’t be used in order to save memory. Or perhaps, I’m misunderstanding how pipelines work. Could you please clarify?

Reply:
If loading the model is very fast, then caching in memory is unnecessary. But my experience has been that sizeable models are not fast to load.

See https://fastapi.tiangolo.com/advanced/events/ for a (maybe) principled solution. Perhaps it should also offload when idle, but it's not even clear that we'd want that (it introduces a failure scenario if peak memory usage across all models exceeds GPU memory).